### PR TITLE
F5 silverline v1 - new integration

### PR DESCRIPTION
**Pull request by abaumgarten**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/20672

## Description
A new integration - F5 Silverline. 
includes:
3 commands:
1. **f5-silverline-ip-objects-list**
2. **f5-silverline-ip-object-add**
3. **f5-silverline-ip-object-delete**
* test-module
* TPB
* UTs
* README

Fetch incidents: with Syslog (server) integration. In order to support this kind of fetch, this pack includes the following as well:
* 4 incident types: F5 Silverline WAF Events, F5 Silverline L7 DDoS Events, F5 Silverline Threat Intelligence Events, F5 Silverline iRule Events.
* 70 incident fields.
* Classifier
* Mapper
* 4 layouts

## Screenshots
For example, 2 incidents:
image

## Minimum version of Cortex XSOAR
- [x] 6.0.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    url_suffix = f'{list_type}/ip_objects'
+
+    body = define_body_for_add_ip_command(list_target, mask, ip_address, duration, note, tags)
+
+    client.request_ip_objects(body=body, method='POST', url_suffix=url_suffix, params={}, resp_type='content')
+    human_readable = f"IP object with IP address: {ip_address} added successfully into the {list_type} list."
+    return CommandResults(readable_output=human_readable)
+
+
+def define_body_for_add_ip_command(list_target, mask, ip_address, duration, note, tags):
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py:98]: Done

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    API docs: https://portal.f5silverline.com/docs/api/v1/ip_objects.md (GET section)
+    """
+    list_type = args['list_type']
+    object_ids = argToList(args.get('object_id'))
+    page_number = args.get('page_number')
+    page_size = args.get('page_size')
+    url_suffix = f'{list_type}/ip_objects'
+    params = {}
+
+    is_paging = False
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py]: Done, changed to `is_paging_required`

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    if page_number and page_size:
+        params = paging_args_to_params(page_size, page_number)
+        is_paging = True
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py]: Done

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        human_results = parse_get_ip_object_list_results(response)
+    else:
+        human_results, outputs = get_ip_objects_by_ids(client, object_ids, list_type, params)  # type: ignore
+
+    human_readable = tableToMarkdown('F5 Silverline IP Objects', human_results, TABLE_HEADERS_GET_OBJECTS,
+                                     removeNull=True)
+
+    if not human_results and is_paging:
+        human_readable = "No results were found. Please try to run the command without page_number and page_size to " \
+                         "get all the IP objects that exist."
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py]: Done

**Comment by abaumgarten** [comment]
@kirbles19 assign you as this is a new integration that requires docs review. thanks!

**Comment by ShirleyDenkberg** [comment]
@guykeller Doc review completed.

**Review by guykeller — COMMENTED** [review]
pending paging demo

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 5]
```diff
@@ -75,6 +75,17 @@ def paging_args_to_params(page_size, page_number):
...
 
 
+def get_ip_and_mask_from_cidr(cidr_range):
+    if '/' not in cidr_range:
+        ip_address = cidr_range
+        mask = '32'
+    else:
+        cidr_range = cidr_range.split('/')
+        ip_address = cidr_range[0]
+        mask = cidr_range[1]
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py:85]: done

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 6]
```diff
@@ -123,19 +123,46 @@ def define_body_for_add_ip_command(list_target, mask, ip_address, duration, note
...
+            return True
+    return False
+
+
 def delete_ip_objects_command(client: Client, args: Dict[str, Any]) -> CommandResults:
     """
-    Deletes an exist IP object from the requested list type (denylist or allowlist) by its object id (mandatory).
-    Note: Human readable appears only if the HTTP request did not fail.
+    Deletes an exist IP object from the requested list type (denylist or allowlist) by its object id or its ip.
+    Note: Human readable appears only if the HTTP request did not fail and the object id exists.
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py:149]: An error will be rasied. I added it to the doc string.

File: Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py [hunk 7]
```diff
@@ -123,19 +123,46 @@ def define_body_for_add_ip_command(list_target, mask, ip_address, duration, note
...
+    object_id = args.get('object_id')
+    object_ip = args.get('object_ip')
+    if not object_ip and not object_ip:
+        raise DemistoException("At least one of the following should be given: object_ip, object_id.")
+
+    if not object_id:
+        object_id = get_object_id_by_ip(client, list_type, object_ip)
+
     url_suffix = f'{list_type}/ip_objects/{object_id}'
+    if is_object_id_exist(client, object_id, list_type):
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/F5Silverline/Integrations/F5Silverline/F5Silverline.py:162]: I think that there is a misunderstanding here.
The API does not return an error when we try to delete an object whose ID does not exist. So to check so, I first check if the id exists and id yes, delete it. If the ID does not exist, the function `get_ip_objects_by_ids` will raise an error and the delete operation will not happen. 
 I did not understand the question about the number of objects, sorry.

**Review by guykeller — APPROVED** [review]

**Pull request merged**